File: Manuel/2025-08-05/수 정렬하기 3/퀵 정렬법/10989.py
import sys
import time

def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    
    # 피봇을 배열의 첫 번쨰 원소로 선택
    pivot = arr[0]
    tail = arr[1:]

    # 피봇보다 작은 원소들의 배열
    left_side = [x for x in tail if x <= pivot]
    # 피봇보다 큰 원소들의 배열
    right_side = [x for x in tail if x > pivot]
    
    # 재귀를 사용하여 정렬
    return quick_sort(left_side) + [pivot] + quick_sort(right_side)

def main():
    sys.setrecursionlimit(10 ** 6)
    input = sys.stdin.readline

    N = int(input())
    numbers = [int(input()) for _ in range(N)]

    start_time = time.perf_counter()

    sorted_numbers = quick_sort(numbers)

    end_time = time.perf_counter()

    for num in sorted_numbers:
        print(num)

    print(f"\n실행 시간: {(end_time - start_time) * 1000:.6f}ms\n", file = sys.stderr)

main()

File: Manuel/2025-08-05/수 정렬하기 3/퀵 정렬법/10989.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <iomanip>
#define endl "\n"

using namespace std;

// 분할 함수
int partition(vector<int>& arr, int left, int right)
{
    // 피봇을 배열의 가장 오른쪽 원소로 선택
    int pivot = arr[right];
    // i는 피봇보다 작은 원소들의 경게선
    int i = left - 1;

    // left부터 right - 1까지 순회하며 피봇과 비교함
    for(int j = left; j < right; j++)
    {
        // 현재 원소가 피봇보다 작거나 같으면:
        if(arr[j] <= pivot)
        {
            // i 증가, i와 j 교환
            i++;
            swap(arr[i], arr[j]);
        }
    }
    // 피봇을 i + 1로 이동시켜 정렬된 위치에 배치
    swap(arr[i + 1], arr[right]);
    // 피봇의 위치 반환
    return i + 1;
}

// 재귀를 활용하여 배열을 정렬하는 퀵 정렬 함수
void quick_sort(vector<int>& arr, int left, int right)
{
    if(left < right)
    {
        // 분할 수행, 피봇 위치 반환
        int pivot_index = partition(arr, left, right);
        
        // 피봇을 기준으로 왼쪽과 오른쪽 각각 배열 정렬
        quick_sort(arr, left, pivot_index - 1);
        quick_sort(arr, pivot_index + 1, right);
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N;
    cin >> N;
    vector<int> numbers(N);
    for(int i = 0; i < N; i++)
    {
        cin >> numbers[i];
    }

    auto start_time = chrono::high_resolution_clock::now();

    quick_sort(numbers, 0, N - 1);

    auto end_time = chrono::high_resolution_clock::now();

    for(int i = 0; i < N; i++)
    {
        cout << numbers[i] << endl;
    }

    chrono::duration<double, milli> duration = end_time - start_time;
    cout << "\n실행 시간: " << fixed << setprecision(6) << duration.count() << "ms" << endl;

    return 0;
}